### ISSUE TITLE
MageOS version

### ISSUE THREAD

**Issue by pjw345** [mage-os/mageos-magento2#108]
### Summary

After migrating to Mage OS 1.0.4 3rd party modules that rely on release version failed as they are checking against Magento release version and then selecting the wrong code version. An example is Mirasvit's Ultimate Search module that checks which Magento version is installed, to then build the correct Class headers, src/SearchElastic/SearchAdapter/Query/Builder/MatchCompatibility.php

as 1.0.4 < 2.4.4 it is selecting the wrong Class headers and the compilation fails.

Do you propose a better was of selecting which version is installed?

### Examples

if (version_compare(CompatibilityService::getVersion(), '2.4.4', '<')) {

### Triage and priority

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

**Comment by fballiano** [comment]
Hi @pjw345, I've wrote on discord to see what's everybody's opinion on this, I hope we'll get some options soon

**Comment by fballiano** [comment]
@rhoerr @sprankhub @aleron75 what's your opinion on this?

**Comment by rhoerr** [comment]
I've run into the same issue with a module from Magic360, where a version check triggered it running incorrect compatibility code.

Since this is a common pattern some vendors seem to use, I think we should handle it on our side if possible.

--

Personally, I like the idea of preserving Magento's (equivalent upstream) version for getVersion(), and then having some separate method to identify the Mage-OS platform and version. That seems like it would be best for intercompatibility.

Actually implementing that will be tricky. Right now I think the only thing mapping Mage-OS 1.0.5 -> Magento 2.4.7-p3 (for instance) is the release build command. Composer.json has our version number. Composer.lock has the individual package versions and replace tags for upstream versions. Probably we need to add something to composer.json metadata during the release generation, for modified code to pull from.

**Comment by fballiano** [comment]
I couldn't find any file with both 1.0.5 and 2.4.7-p3 on https://github.com/mage-os/generate-mirror-repo-js/ :-(

**Comment by rhoerr** [comment]
> I couldn't find any file with both 1.0.5 and 2.4.7-p3 on https://github.com/mage-os/generate-mirror-repo-js/ :-(

Yeah, that's (sort of) part of what we need to fix. The build version and upstream version for the next release are only defined when executing the workflow at https://github.com/mage-os/generate-mirror-repo-js/actions/workflows/build-mageos-release.yml (`mageos_release`, `upstream_release`). We'll need to modify that to incorporate them into the generated base json, or similar.

The path here is
https://github.com/mage-os/generate-mirror-repo-js/blob/main/.github/workflows/build-mageos-release.yml
-> https://github.com/mage-os/generate-mirror-repo-js/blob/main/.github/workflows/deploy.yml
-> https://github.com/mage-os/generate-mirror-repo-js/blob/main/src/make/mageos-release.js
-> https://github.com/mage-os/generate-mirror-repo-js/blob/main/src/release-build-tools.js#L314
-> Whichever specific part builds the main composer.json -- at a glance, possibly:?
 https://github.com/mage-os/generate-mirror-repo-js/blob/main/src/release-build-tools.js#L218
-> https://github.com/mage-os/generate-mirror-repo-js/blob/main/src/package-modules.js#L523

**Comment by fballiano** [comment]
I tried to read a bit of this process but here the "problem" is in the fact that all versions are read from composer and since mage-os "replaces" magento it automatically replaces magento's version, so either we introduce a meta package representing the magento version and then we have a workflow to keep it up to date, or we hardcode it somewhere.

but there's still the problem of having the core functions returning mage-os version instead of magento version...

the only way I see it is that some senior tech responsible for mage-os decides a way to go with this and implement it.

**Comment by rhoerr** [comment]
Picking this up to give it a try.

Based on https://getcomposer.org/doc/04-schema.md#extra, it should be possible to define extra keys in the base composer.json to record the marketing version of Mage-OS and upstream. Then it's a matter of pulling those in the version class, instead of the current data source.

**Comment by fballiano** [comment]
This should be fixed now, since 1.0.6 or 1.1

**Comment by pjw345** [comment]
Actually I think it is broken.  On version 1.0.6 it now returns UNKNOWN which makes things worse when 3rd party developers use list($a, $b) after calling getVersion().  I partly blame the 3rd party developers for not checking to see if a version is indeed returned and this implementation.. If it cannot return the Magento version, then it should at least return the MageOS version.

**Comment by rhoerr** [comment]
Hi @pjw345 -- could you share your composer.json and composer.lock files? Thanks!

**Comment by pjw345** [comment]
I think the problem is around the function isMagentoRoot() as the $rootPackage currently is "mage-os/project-community-edition-1.0.6" and hence does not return true, on the preg_match mage-os/magento2

With regards to composer.json and composer.lock, I've added magento_version to the extra section of the composer.json and appears correctly in the composer.lock file, so I don't think that's the issue.
